Test textgrid

keggsmurph21 opened the pull request:
@calebho Do you have an idea of how you'd want the `AlignmentFileLoader` API to work?  I wrote a bare-bones interface [here](https://github.com/SwatPhonLab/UltraTrace/blob/04717fdba42c2732a161748879aae56e3628bea0/ultratrace2/model/files/loaders/base.py).

keggsmurph21 committed:
TextGridLoader: Add tests

keggsmurph21 committed:
Merge branch 'master' into test-textgrid

keggsmurph21 committed:
DICOMLoader: Fix f-string typo

keggsmurph21 committed:
TextGridLoader: Add very basic file-loading + tests

keggsmurph21 committed:
AlignmentFileLoader: add some abstract methods & subclass impls

Added the following abstract methods to the base class:
 - get_tier_names
 - get_start
 - get_end
 - get_offset
 - set_offset

For each of these methods, implemented TextGridLoader::$method and
stubbed MeasurementLoader::$method.

Also added some basic tests for each of these methods :^)

keggsmurph21 committed:
DICOMLoader: Remove extraneous print() call :^)

keggsmurph21 committed:
AlignmentFileLoader: Add an abstract ::get_intervals() method

This method should return a Sequence of (name, Sequence[Interval])
tuples, where each interval corresponds to one "drawable" alignment
interval.

The first axis corresponds to what we would think of as "vertical"
differences -- the different tiers in TextGridWorld -- and the second
axis corresponds to "horizontal" -- the different intervals of a
specific tier in TextGridWorld.  The reason we iterate Tuples instead
of just using, say, an OrderedDict, is because the tier names can have
duplicates.

This patch also implements some tests :^)

keggsmurph21 committed:
TextGridLoader: Correctly parse ASCII, Latin-1, Mac-Roman encodings

This patch also includes some example textgrid files adapted from the
Praat man page on File Formats[1].  According to that page, the textgrid
library we're using does not completely follow the spec (it assumes we
are always giving it human-readable input).  However, it does detail
some "expected" encodings, so those are implemented in this patch.

[1] http://www.fon.hum.uva.nl/praat/manual/TextGrid_file_formats.html

calebho commented on ultratrace2/model/files/loaders/tests/test_textgrid.py:
```diff
@@ -0,0 +1,936 @@
+        TextGridLoader.from_file(path)
+
+
+@pytest.mark.parametrize(
```
😍

calebho commented on ultratrace2/model/files/loaders/textgrid.py:
```diff
@@ -8,9 +35,86 @@ def get_path(self) -> str:
+            transcoded_file = cls.copy_to_temp_file_with_encoding(path, encoding)
+            return textgrid.TextGrid.fromFile(transcoded_file.name)
+
+    @classmethod
```
This can be `staticmethod`

keggsmurph21 replied:
fcc430a

calebho commented on ultratrace2/model/files/loaders/textgrid.py:
```diff
@@ -8,9 +35,86 @@ def get_path(self) -> str:
+                f"Invalid TextGrid ({path}), unable to read: {str(e)}"
+            ) from e
+
+    @classmethod
```
This can also be `staticmethod` (see below)

keggsmurph21 replied:
fcc430a

calebho commented on ultratrace2/model/files/loaders/textgrid.py:
```diff
@@ -8,9 +35,86 @@ def get_path(self) -> str:
+    def copy_to_temp_file_with_encoding(
+        cls, original_path: str, encoding: str
+    ) -> IO[bytes]:
+        # this gets ::close()d when it is GC'ed
```
Ditto. Alternatively, you can use it as a context manager and have `close()` called when the context is exited rather than relying on GC:
```
with tempfile.NamedTemporaryFile as temp_file:
    ...
```
Though you'd have to move this `with` to the call site  and pass the temporary file as an argument.

keggsmurph21 replied:
5eff9ae

calebho reviewed (approved):
lgtm

keggsmurph21 committed:
TextGridLoader: Make @classmethods @staticmethods (#118)

keggsmurph21 committed:
TextGridLoader: Explicitly manage our temporary file contexts

Before, we were relying on garbage collection to clean up our temp
files, which is probably not tenable long-term.
